### DISCUSSION TITLE
pozzo universale

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#8] [Esercizi A-L]
Dato un grafo diretto $G = (V, E)$ un **pozzo universale** è un vertice $x \in V$ tale che
- $\nexists \\; y \in V \mid (x, y) \in E$ (non ha archi uscenti)
- $\forall y \in V : y \neq x \implies (y, x) \in E$ (ogni altro vertice $y$ ha un arco che entra in $x$)

Dato $G$ come **matrice di adiacenza** scrivere lo pseudocodice di un algoritmo che verifica se $G$ ha un pozzo universale in $O(|V|)$

**Comment by Frisbon** [comment]
### Se avete da contraddirmi, fatelo perfavore, sul serio

Suppongo che l’algoritmo del professore _(preso dagli appunti di SimLid)_ si basa su alcune osservazioni, ossia che nella matrice di adiacenza, un pozzo _(vedi grafo grigio)_ viene riconosciuto se: 
- la sua riga é tutta 0 
- la sua colonna é tutta 1 _(tranne dove i=j)_.

> Questo é **falso**. 

Come visto in figura (grafo verde), per definizione, 2 é un pozzo e l’algoritmo applicato su questa matrice di adiacenza mi ritorna False.

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/84731341/e75c1021-702f-48f4-8428-aad63ab14c6e)

Ció avviene perché l’algoritmo non si aspetta che gli altri nodi puntino a qualcos’altro oltre al pozzo. *(Si aspetta il grafo grigio)*

Una soluzione alternativa che (purtroppo) é in $O(n^2)$:

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/84731341/e78d9eae-49cf-4ed3-9d01-76f23b9d3aeb)

**Comment by LeoCosta02** [comment]
pozzo_universale(M) // input matrice di adiacenza 

candidato=1
n=m[1].length() // n è il numero di vertici

 for(i=2; i<=n;i++){  // se esiste un arco tra il candidato ed il nodo j allora cambio candidato 
  if(m[candidato,i]==1)
    candidato=i
 }

   for(i=1;i<=n;i++){ //verifico se il candidato non ha archi verso altri nodi 
    if(m[candidato,i]==1)
      return false

   }

     for(i=1,i<=n;i++){ //controllo se tutti i nodi hanno archi verso il candidato
      if(m[i,candidato]==0 and i!=candidato)
        return false

     }

 return true

**Comment by niccolo-fato** [comment]
Come prima cosa cerco  un buon candidato come pozzo universale:

```python
Function(m): #Matrice in input
	candidato = 1 #Parto dal primo nodo come candidato
	n = m[1].lenght #Numeri di vertici presenti nel grafo(5 in questo caso)
	for i=2; i≤n; i++: #Controllo i vertici dal secondo in poi
		if m[candidato, i]==1:
			candidato = i
```
![Diagramma_senza_titolo drawio 24](https://github.com/user-attachments/assets/3801b57f-16f0-4bc6-b2b1-172759336003)
Successivamente controllo se il candidato ha un arco uscente, se li ha allora il grafo non ha un pozzo universale:

```python
for i=1; i≤n; i++:
	if m[candidato, i]==1:
		return False #Il candidato ha un arco uscente
```
In fine controllo che tutti i vertici(escluso il candidato) abbiano un arco verso il candidato:

```python
for i=1; i≤n; i++:
	if m[i, candidato]==0 and i!=candidato:
		return False #Almeno un vertice non ha un arco verso candidato
return True
```
